src/pyggdrasil/tree_inference/_mcmc_sampler.py

```diff
@@ -0,0 +1,164 @@
+
+    # define loop body
+    def body(
+        state: Tuple[int, JAXRandomKey, Array, Array, float],
+    ) -> Tuple[int, JAXRandomKey, Array, Array, float]:
+        """Body of the MCMC loop.
+
+        Args:
+            state: tuple containing the current state of the MCMC sampler
+                - iteration: current iteration
+                - rng_key: random key for the MCMC sampler
+                - Array: current tree, adjacency matrix
+                - Array: labels of the tree
+                - float: log-probability of the current tree
+
+        Returns:
+                 updated state: tuple containing the current state of the MCMC sampler
+                - iteration: current iteration
+                - rng_key: random key for the MCMC sampler
+                - Array: current tree, adjacency matrix
+                - Array: labels of the tree
+                - float: log-probability of the current tree
+        """
+
+        # get current state
+        iter_sample, rng_key_body, topo, labels, logprobability = state
+        iter_sample = iter_sample + 1
+
+        # make Tree
+        tree = Tree(tree_topology=topo, labels=labels)
+
+        # split random key to use one
+        rng_key_run, rng_key_sample = jax.random.split(rng_key_body)
+
+        # mcmc kernel
+        tree, logprobability = mcmc._mcmc_kernel(
+            rng_key_sample,
+            tree,
+            data,
+            error_rates,
+            move_probs,
+            logprobability_fn,
+            logprobability,
+        )
+        logging.info("Iteration: %d, log-probability: %f", iter_sample, logprobability)
+
+        # burn-in - do not save samples in burning phase
+        if iter_sample > num_burn_in:
+            # save sample
+            if iter_sample % thinning == 0:
+                # pack sample
+                sample = mcmc_util._pack_sample(iter_sample, tree, logprobability)
+                # save sample
+                serialize.save_mcmc_sample(
+                    sample, output_dir, timestamp=kwargs["timestamp"]
+                )
+                logging.info("Saved sample %d.", iter_sample)
+
+        # return updated state
+        topo = tree.tree_topology
+        labels = tree.labels
+        updated_state = (iter_sample, rng_key_run, topo, labels, logprobability)
+
+        return updated_state
+
+    # conditional function for MCMC kernel
+    def cond_fn(state):
+        """Condition for the MCMC loop."""
+        iter_sample, _, _, _, _ = state
+        return jax.lax.lt(iter_sample, num_samples)
+
+    # mcmc loop
+    # TODO: use jax.lax.while_loop instead of while_loop,
+    # requires further modification of _get_descendants at least
+    # jax.lax.while_loop(cond_fn, body, init_state)
+
+    def while_loop(cond_fun, body_fun, init_val):
+        """While loop for MCMC sampler."""
+        val = init_val
+        while cond_fun(val):
+            val = body_fun(val)
+        return val
```

Comment:
Wrote this function in a way, to be usable with `jax.lax.while_loop`. 

Some modifications are needed in the `mcmc kernel` and in deeper down in `get_descendants`. Might need some help there. - Couldn't make sense of some `jax._src.errors.ConcretizationTypeError`.

Reply:
I think it currently has the problem of tracing the shapes of arrays. One reason for this could be if e.g., number of mutations (as an integer) is passed in the function argument.

Reply:
Ok, hope we get to this at some stage. Interested to see how fast JAX is in relation.